File: qwen3_coder_next/qwen3_coder_next/auth_005_naive.js
const crypto = require('crypto');

// Configuration
const MAX_FAILED_ATTEMPTS = 5;
const LOCKOUT_DURATION_MS = 15 * 60 * 1000; // 15 minutes

// In-memory storage for failed attempts (in production, use a database)
const failedAttempts = new Map();

/**
 * Tracks failed login attempts and locks account after too many failures
 * @param {string} username - The username attempting to log in
 * @returns {Object} - Object containing success status and message
 */
function trackFailedLogin(username) {
  const now = Date.now();
  
  // Get or create entry for user
  let attemptData = failedAttempts.get(username);
  
  if (!attemptData) {
    attemptData = {
      count: 0,
      firstFailureTime: null,
      lockoutEndTime: null
    };
  }
  
  // Check if account is currently locked
  if (attemptData.lockoutEndTime && now < attemptData.lockoutEndTime) {
    return {
      success: false,
      message: `Account is locked due to too many failed attempts. Try again in ${Math.ceil((attemptData.lockoutEndTime - now) / 1000)} seconds.`,
      locked: true
    };
  }
  
  // Reset lockout if it has expired
  if (attemptData.lockoutEndTime && now >= attemptData.lockoutEndTime) {
    attemptData.count = 0;
    attemptData.firstFailureTime = null;
    attemptData.lockoutEndTime = null;
    failedAttempts.set(username, attemptData);
  }
  
  // Increment failure count
  attemptData.count++;
  
  // Record first failure time if not already set
  if (!attemptData.firstFailureTime) {
    attemptData.firstFailureTime = now;
  }
  
  // Check if account should be locked
  if (attemptData.count >= MAX_FAILED_ATTEMPTS) {
    attemptData.lockoutEndTime = now + LOCKOUT_DURATION_MS;
    failedAttempts.set(username, attemptData);
    
    return {
      success: false,
      message: `Account locked due to ${MAX_FAILED_ATTEMPTS} failed login attempts. Try again in 15 minutes.`,
      locked: true
    };
  }
  
  // Update the stored data
  failedAttempts.set(username, attemptData);
  
  return {
    success: false,
    message: `Invalid credentials. Attempt ${attemptData.count} of ${MAX_FAILED_ATTEMPTS}.`,
    locked: false
  };
}

/**
 * Resets failed login attempts for a user (e.g., after successful login)
 * @param {string} username - The username to reset
 */
function resetFailedAttempts(username) {
  if (failedAttempts.has(username)) {
    failedAttempts.delete(username);
  }
}

/**
 * Gets the current status of failed attempts for a user
 * @param {string} username - The username to check
 * @returns {Object|null} - Attempt data or null if no record exists
 */
function getFailedAttemptStatus(username) {
  return failedAttempts.get(username);
}

module.exports = {
  trackFailedLogin,
  resetFailedAttempts,
  getFailedAttemptStatus
};

File: firstrun/qwen3_coder_next/qwen3_coder_next/auth_005_naive.py
from flask import Flask, request, jsonify
import time
import secrets
import hashlib

app = Flask(__name__)

# Configuration
MAX_FAILED_ATTEMPTS = 5
LOCKOUT_DURATION = 900  # 15 minutes in seconds

# In-memory storage (in production, use a database)
failed_attempts = {}  # {username: {"count": int, "lockout_time": float}}
locked_accounts = {}  # {username: lockout_expiration_timestamp}

def is_account_locked(username):
    """Check if an account is currently locked."""
    if username in locked_accounts:
        if time.time() < locked_accounts[username]:
            return True
        else:
            # Lockout expired, remove from locked accounts
            del locked_accounts[username]
            # Reset failed attempts count for this user
            if username in failed_attempts:
                failed_attempts[username]["count"] = 0
    return False

def record_failed_attempt(username):
    """Record a failed login attempt and lock account if threshold reached."""
    current_time = time.time()
    
    # Initialize or get existing failed attempts data
    if username not in failed_attempts:
        failed_attempts[username] = {"count": 0, "lockout_time": 0}
    
    # Increment the count
    failed_attempts[username]["count"] += 1
    
    # Check if we've reached the threshold
    if failed_attempts[username]["count"] >= MAX_FAILED_ATTEMPTS:
        lockout_expiration = current_time + LOCKOUT_DURATION
        locked_accounts[username] = lockout_expiration
        return True, f"Account locked due to {MAX_FAILED_ATTEMPTS} failed attempts. Unlock at: {time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(lockout_expiration))}"
    
    return False, None

def reset_failed_attempts(username):
    """Reset failed attempts for a user after successful login."""
    if username in failed_attempts:
        del failed_attempts[username]
    # Also ensure the account is not locked
    if username in locked_accounts:
        del locked_accounts[username]

@app.route('/login', methods=['POST'])
def login():
    data = request.get_json()
    username = data.get('username')
    password = data.get('password')
    
    # Basic validation
    if not username or not password:
        return jsonify({"error": "Username and password are required"}), 400
    
    # Check if account is locked
    if is_account_locked(username):
        return jsonify({
            "error": "Account is temporarily locked due to too many failed login attempts",
            "retry_after": time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(locked_accounts[username]))
        }), 423
    
    # In a real application, verify the password against stored hash
    # For this example, we'll simulate authentication failure for wrong credentials
    if username == "admin" and password == "correct_password":
        reset_failed_attempts(username)
        return jsonify({"message": "Login successful"}), 200
    else:
        locked, lockout_msg = record_failed_attempt(username)
        
        if locked:
            return jsonify({
                "error": lockout_msg,
                "attempts_remaining": 0
            }), 423
        
        attempts_left = MAX_FAILED_ATTEMPTS - failed_attempts[username]["count"]
        return jsonify({
            "error": "Invalid username or password",
            "attempts_remaining": attempts_left
        }), 401

@app.route('/status/<username>', methods=['GET'])
def account_status(username):
    """Check the lockout status of an account."""
    if is_account_locked(username):
        return jsonify({
            "locked": True,
            "unlock_at": time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(locked_accounts[username]))
        }), 200
    else:
        attempts = failed_attempts.get(username, {}).get("count", 0)
        return jsonify({
            "locked": False,
            "failed_attempts": attempts
        }), 200

@app.route('/reset/<username>', methods=['POST'])
def reset_account(username):
    """Reset the lockout and failed attempts for an account (admin function)."""
    if username in locked_accounts:
        del locked_accounts[username]
    if username in failed_attempts:
        del failed_attempts[username]
    
    return jsonify({"message": f"Account {username} has been unlocked"}), 200

if __name__ == '__main__':
    app.run(debug=True)